pulp_container/tests/functional/api/test_push_content.py

```diff
@@ -6,49 +6,145 @@
+    PulpContainerNamespacesApi,
+    RepositoriesContainerPushApi,
+)
 
 
-cfg = config.get_config()
+class PushRepoTestCase(unittest.TestCase):
+    """Verify whether images can be pushed to pulp."""
 
-api_client = gen_container_client()
-distributions_api = DistributionsContainerApi(api_client)
+    @classmethod
+    def setUpClass(cls):
+        """
+        Define APIs to use and pull images needed later in tests
+        """
+        api_client = gen_container_client()
+        cls.pushrepository_api = RepositoriesContainerPushApi(api_client)
+        cls.namespace_api = PulpContainerNamespacesApi(api_client)
 
-registry = cli.RegistryClient(cfg)
+        cfg = config.get_config()
+        cls.registry = cli.RegistryClient(cfg)
+        cls.registry.raise_if_unsupported(unittest.SkipTest, "Tests require podman/docker")
+        cls.registry_name = urlparse(cfg.get_base_url()).netloc
 
+        admin_user, admin_password = cfg.pulp_auth
+        cls.user_admin = {"username": admin_user, "password": admin_password}
+        cls.user_creator = gen_user(
+            [
+                "container.add_containerdistribution",
+                "container.view_containerdistribution",
+                "container.add_containernamespace",
+            ]
+        )
+        cls.user_reader = gen_user(["container.view_containerpushrepository"])
+        cls.user_helpless = gen_user([])
 
-class PushContentTestCase(unittest.TestCase):
-    """Verify whether images can be pushed to pulp."""
+        cls._pull("centos:7")
+        cls._pull(f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a")
+
+    @classmethod
+    def tearDownClass(cls):
+        """Delete api users."""
+        del_user(cls.user_creator)
+        del_user(cls.user_reader)
+        del_user(cls.user_helpless)
+
+    @classmethod
+    def _pull(cls, image_path, user=None):
+        """
+        Pull using specified user.
+        """
+        if user:
+            cls.registry.login("-u", user["username"], "-p", user["password"], cls.registry_name)
+        else:
+            # Ensure logout
+            try:
+                cls.registry.logout(cls.registry_name)
+            except exceptions.CalledProcessError:
+                pass
+
+        cls.registry.pull(image_path)
+
+        if user:
+            cls.registry.logout(cls.registry_name)
 
-    def test_push_using_registry_client(self):
-        """Test push with official registry client"""
-        registry.raise_if_unsupported(unittest.SkipTest, "Test requires podman/docker")
-        registry_name = urlparse(cfg.get_base_url()).netloc
-        local_url = "/".join([registry_name, "foo/bar:1.0"])
-        # Be sure to not being logged in
-        try:
-            registry.logout(registry_name)
-        except exceptions.CalledProcessError:
-            pass
-        # Pull an image with large blobs
-        registry.pull("centos:7")
+    @classmethod
+    def tag_and_push(cls, image_path, local_url, user):
```

Comment:
@mdellweg , is it any better? or would you prefer to copy-paste these 5 lines of code + comments to every test?

Reply:
Sorry, i was mistaken by the `registry.rmi` call. So to get it straight: This function tags a local image to the registry, logs in pushes the image and removes it from the _local_ storage. That seems like a nice primitive in tests.

I thought that it also removed it from the registry right away, and it didn't occur to me how we could test pulling it with another user afterwards. So maybe `_push` was a good name.

If you want to start testing that a push operation can be denied, you can wrap it in a `try` `finally` block.
(suggestions below)